### PR TITLE
attempt at chat function

### PR DESCRIPTION
**Pull request by karpathy**
just uploading partial work of the chat function, version 1
not sure i'm super happy with the layout of it just yet
weird hard-coded buffer sizes
seems to not be obviously broken though, e.g. on my machine:

```
ubuntu:~/llamac$ make runomp
ubuntu:~/llamac$ OMP_NUM_THREADS=16 ./run llama2_7b_chat.bin -m chat -y "You only respond with just the first single word that comes to mind" -i "America"

<s>
[INST] <<SYS>>
You only respond with just the first single word that comes to mind
<</SYS>>

America
[/INST]  Flag
</s>
Enter user prompt: Canada

<s>
[INST] Canada [/INST]  Maple
</s>
Enter user prompt: France

<s>
[INST] France [/INST]  Eiffel
</s>
Enter user prompt: Italy

<s>
[INST] Italy [/INST]  Pizza
</s>
Enter user prompt: Iceland

<s>
[INST] Iceland [/INST]  Glacier
</s>
Enter user prompt: Middle Earth

<s>
[INST] Middle Earth [/INST]  Hobbit
</s>
Enter user prompt: 
```

### CODE REVIEW

**Comment by karpathy** [comment]
Once I validate that all the special tokens and the schema is correctly followed, I think it would make sense to hide all that stuff away and make it a pretty chat.

**Comment by karpathy** [comment]
Another thought: it *might* be possible to re-use the `generate` function for chat. but it would require quite a few changes inside it, not sure if that's the better path. Leaning to no atm.

**Comment by rdentato** [comment]
I think is important to avoid having to reload the model every time.
With CUDA (and, in general, using a GPU) it will be a slow process. It takes almost 2min on my T4 to load the fp16 llama2-7b-chat model.

**Comment by karpathy** [comment]
> I think is important to avoid having to reload the model every time.
> With CUDA (and, in general, using a GPU) it will be a slow process. It takes almost 2min on my T4 to load the fp16 llama2-7b-chat model.

is this comment related to this PR?

**Comment by rdentato** [comment]
Yes, but neither I remember exactly what I wanted to say. Sorry.

**Comment by rdentato** [comment]
Anyway, my biggest concern on this approach is that it has the template used by llama_7_b_chat embedded into the code.
I don't think it should be merged into the master until there's a way to choose the template.
Unless you consider this version temporary and plan to add the ability to use different templates to a later version.

**Comment by karpathy** [comment]
What other templates are there though? Isn't this _the_ template for Llama 2 chat?

**Comment by rdentato** [comment]
What if anyone trains another chat model using its own template? Not necessarily from one of the Meta lama models.
Isn't llama2.c intent to allow people to train (or fine tune) their model and then launch inference on them?

**Comment by karpathy** [comment]
If they use the exact same format as Llama 2 or finetune a Llama 2 chat model then the code is plug and play
If they wish to innovate on their format then they have to develop their own chat code

**Comment by rdentato** [comment]
Ok. I see your point. I still believe it should be made more general but it's something that could wait for when someone will want to introduce a new template.

File: run.c [hunk 1]
```diff
@@ -785,6 +787,108 @@ void generate(Transformer *transformer, Tokenizer *tokenizer, Sampler *sampler,
...
+            token = prompt_tokens[user_idx++];
+        } else {
+            // otherwise use the next token sampled from previous turn
+            token = next;
+        }
+        // EOS (=2) token ends the Assistant turn
+        if (token == 2) { user_turn = 1; }
+
+        // forward the transformer to get logits for the next token
+        float* logits = forward(transformer, token, pos);
```
Comment (nikolaydubina) [inline run.c:875]: just for my knowledge, isn't there any problems with not forwarding model through tokens from users prompt? (e.g. things like KV cache and offsets?)

**Pull request merged**